File: dsa_with_python/tree/binary_search_tree/bst_iterative.py
from node import Node

root = None


def insert(data):
    global root

    def __insert(root: Node, data) -> Node:
        new_node = Node(data)
        parent_node, current_node = None, root
        while current_node is not None:
            parent_node = current_node
            if current_node.data > data:
                current_node = current_node.left
            elif current_node.data < data:
                current_node = current_node.right
            else:
                return root
        if parent_node is None:
            return new_node
        if parent_node.data > data:
            parent_node.left = new_node
        else:
            parent_node.right = new_node
        return root

    root = __insert(root, data)


def search(data) -> bool:
    global root

    def __search(root: Node, data) -> bool:
        while root is not None:
            if root.data == data:
                return True
            elif root.data > data:
                root = root.left
            else:
                root = root.right
        return False

    return __search(root, data)


def display():
    def __display(root: Node):
        stack = []
        current_node = root
        print("[", end="")
        while current_node is not None or len(stack) > 0:
            while current_node is not None:
                stack.append(current_node)
                current_node = current_node.left
            current_node = stack.pop()
            print(current_node.data, end=", ")
            current_node = current_node.right
        print("\b\b]")

    __display(root)

File: DsaWithCpp/Tree/BinarySearchTree/BSTIterative.cpp
#include <iostream>
#include <stack>
#include "Node.cpp"
using namespace std;

class BSTIterative
{
    Node *root = NULL;

public:
    void insert(int data)
    {
        root = insert(root, data);
    }
    bool search(int data)
    {
        return search(root, data);
    }
    void display()
    {
        display(root);
        cout << endl;
    }

private:
    Node *insert(Node *root, int x)
    {
        Node *newNode = new Node(x);
        Node *parentNode = NULL, *currentNode = root;
        while (currentNode != NULL)
        {
            parentNode = currentNode;
            if (currentNode->data > x)
                currentNode = currentNode->left;
            else if (currentNode->data < x)
                currentNode = currentNode->right;
            else
                return root;
        }
        if (parentNode == NULL)
            return newNode;
        if (parentNode->data > x)
            parentNode->left = newNode;
        else
            parentNode->right = newNode;
        return root;
    }
    bool search(Node *root, int x)
    {
        while (root != NULL)
        {
            if (root->data == x)
                return true;
            else if (root->data > x)
                root = root->left;
            else
                root = root->right;
        }
        return false;
    }

    void display(Node *root)
    {
        stack<Node *> stk;
        Node *currentNode = root;
        cout << "[";
        while (currentNode != NULL || stk.empty() == false)
        {
            while (currentNode != NULL)
            {
                stk.push(currentNode);
                currentNode = currentNode->left;
            }
            currentNode = stk.top();
            stk.pop();
            cout << currentNode->data << ", ";
            currentNode = currentNode->right;
        }
        cout << "\b\b]\n";
    }
};